crates/router/src/core/user.rs

```diff
@@ -1320,6 +1321,177 @@ pub async fn get_user_details_in_merchant_account(
+                } else {
+                    None
+                }
+            }
+        };
+
+        let profile = match entity_type.ok_or(UserErrors::InternalServerError)? {
+            EntityType::Internal => {
+                return Err(UserErrors::InvalidRoleOperationWithMessage(
+                    "Internal roles are not allowed for this operation".to_string(),
+                )
+                .into());
+            }
+            EntityType::Organization | EntityType::Merchant => None,
+            EntityType::Profile => {
+                if let Some(profile_id) = &user_role.profile_id {
+                    let key_manager_state = &(&state).into();
+                    let merchant_key_store = state
+                        .store
+                        .get_merchant_key_store_by_merchant_id(
+                            key_manager_state,
+                            &user_role
+                                .merchant_id
+                                .clone()
+                                .ok_or(UserErrors::InternalServerError)?,
+                            &state.store.get_master_key().to_vec().into(),
+                        )
+                        .await
+                        .change_context(UserErrors::InternalServerError)
+                        .attach_printable("Failed to retrieve merchant key store by merchant_id")?;
+
+                    let profile_name = state
+                        .store
+                        .find_business_profile_by_profile_id(
+                            key_manager_state,
+                            &merchant_key_store,
+                            profile_id,
+                        )
+                        .await
+                        .change_context(UserErrors::InternalServerError)
+                        .attach_printable("Business profile not found")?
+                        .profile_name;
+
+                    Some(NameIdUnit {
+                        id: profile_id.clone(),
+                        name: profile_name,
+                    })
+                } else {
+                    None
+                }
+            }
+        };
+
+        let role_info = if let Some(role) =
+            roles::predefined_roles::PREDEFINED_ROLES.get(&user_role.role_id.as_str())
+        {
+            role.clone()
+        } else {
+            state
+                .store
+                .find_role_by_role_id_in_merchant_scope(
+                    &user_role.role_id,
+                    &user_role
+                        .merchant_id
+                        .ok_or(UserErrors::InternalServerError)
+                        .attach_printable("Merchant id not found for custom role")?,
+                    &user_from_token.org_id,
+                )
+                .await
+                .change_context(UserErrors::InternalServerError)
+                .attach_printable("Failed to get custom role")?
+                .into()
+        };
```

Comment:
Why not `from_role_id()`?

Reply:
coz merchant id is non option param in that, he we may or may not get it.

Reply:
Make a function in roles which takes user role and gives back a role info.

If the role_id is present in the predefined roles it will not do any db calls.
Else if the entity type is org, then throw internal server error (As we don't have any org level custom roles.)
Else if the entity type is merchant, then get merchant id from the user role (if it is not there throw 500) and find the custom role.
Else if the entity type is merchant or profile, then get merchant id from the user role (if it is not there throw 500) and find the custom role.

You can change the `from_role_id()` function if it is possible.